### ISSUE TITLE
API: make first argument to read_.. functions positional-only and use consistent name

### ISSUE THREAD

**Issue by jorisvandenbossche** [pandas-dev/pandas#54616]
Currently many of the `read_..` functions have some inconsistent name. We would like to make this consistent, and at the same time make them positional only keywords (we think no one should ever do `read_csv(filepath_or_buffer="path/to/csv")` with an explicit first keyword). 

We can do this without breaking code by:

- adding a new positional only first keyword
- moving the existing first keyword to the end of the named keywords, so we leave it working if someone specified it (and we can also deprecate that later)

The "big" question is which name to use for the positional-only keyword. Some of the existing names we use:

- `filepath_or_buffer` (csv, pickle, sas, stata)
- `io` (excel, html)
- `path_or_buf` (json, xml, hdf)
- `path` (feather, parquet, orc, spss)

Other options we don't currently use:

- `file`
- `input_file`
- `source`
- ...?

Any strong preferences for which name to use?

**Comment by rhshadrach** [comment]
I'm against `io` because of the name collision. I like `path` because it is short and `filepath_or_buffer` because it is more accurate. `path_or_buf` seems like a nice compromise.

**Comment by jorisvandenbossche** [comment]
The "buffer" is also not completely correct, since we also support file-like object that are not necessarily in-memory buffers already (for me a reason to not care too much about having "_buf(fer)" in the name, but also fine with having it)

I might like `input_file` because it sounds more generic (it doesn't explicitly have the meaning of a path or a buffer)

**Comment by asishm-wk** [comment]
are we only talking about file io `read_...` methods here? 

1. `read_sql`, `read_sql_query`: `sql`
2. `read_sql_table`: `table`
3. `read_gbq`: `query`
4. `read_clipboard` shouldn't have a default

**Comment by rmhowe425** [comment]
take

**Comment by jorisvandenbossche** [comment]
> are we only considering file io based `read_...` methods here?

Yes, I didn't list the sql and gbq ones because I felt that their first keywords had a meaningful name (with a good reason to deviate, and thus not needed to make consistent with the others)

**Comment by rhshadrach** [comment]
I'm fine to go along with `input_file` (or really any other ones), but given @jorisvandenbossche critique of "buffer" I find myself now preferring `source`. It seems to work well whether using a path, URI, or buffer.

**Comment by jorisvandenbossche** [comment]
I am fine with `source` as well (although then we will also need to find another name for all `to ..` methods, like sink or destination)

cc @twoertwein

**Comment by twoertwein** [comment]
I have no strong opinion how to name an argument that literally cannot be used by any user (positional-only) :)

I agree with @jorisvandenbossche that it would be nice to have one consistent name between read/to methods. (One could also argue that the names don't need to be consistent, even within read/to, as those names are only relevant to internal pandas code but not to users).

**Comment by jorisvandenbossche** [comment]
We still need to choose something, we can't just keep the inconsistent current names (even though it wouldn't matter much if you can't specify them when they are positional only), as we need to use another name to be able to migrate to positional only.

So if there are no strong opinions here, let's go with `source` for the read methods, and we can still think of something else for the write methods.

**Comment by rhshadrach** [comment]
> will also need to find another name for all `to ..` methods, like sink or destination

As much as the graph theorist in me likes "sink", I think destination would be a bit more clear for the typical user.

+1 on source and destination.
